File: web_sdk/src/crypto/ratchet-engine.ts
// src/crypto/ratchet-engine.ts

import type { RatchetSession, EncryptResult, DecryptResult, RatchetHeader } from './crypto-types.js';
import {
  generateX25519KeyPair,
  exportPublicKeyRaw,
  exportPrivateKeyPkcs8,
  importX25519PublicKey,
  importX25519PrivateKey,
  x25519Dh,
  hkdf,
  aesGcmEncrypt,
  aesGcmDecrypt,
} from './crypto-utils.js';
import { HKDF_INFO } from './crypto-utils.js';
import { base64urlEncode, base64urlDecode } from '../utils/encoding.js';
import { SdkError, SdkErrorCode } from '../errors/sdk-error.js';
import type { RatchetState } from '../session/session-store.js';

const MAX_SKIP = 2000;

/**
 * Double Ratchet Algorithm engine providing forward secrecy and break-in recovery.
 */
export class RatchetEngine {
  /**
   * Initialize a ratchet session from the sender's (initiator) side.
   */
  static async initSender(
    sharedSecret: Uint8Array,
    recipientDhPubBytes: Uint8Array,
  ): Promise<RatchetSession> {
    const dhKp = await generateX25519KeyPair();
    const dhPub = await exportPublicKeyRaw(dhKp.publicKey);
    const dhPriv = await exportPrivateKeyPkcs8(dhKp.privateKey);

    const recipientDhPub = await importX25519PublicKey(recipientDhPubBytes);
    const dhOut = await x25519Dh(dhKp.privateKey, recipientDhPub);
    const [newRootKey, chainKeySend] = await ratchetKdf(sharedSecret, dhOut);

    return {
      rootKey: newRootKey,
      chainKeySend,
      chainKeyRecv: new Uint8Array(32),
      dhSendPub: dhPub,
      dhSendPriv: dhPriv,
      dhRecvPub: recipientDhPubBytes,
      nSend: 0,
      nRecv: 0,
      pn: 0,
      skippedMessageKeys: new Map(),
    };
  }

  /**
   * Initialize a ratchet session from the receiver's (responder) side.
   */
  static async initReceiver(
    sharedSecret: Uint8Array,
    localDhKeyPair: { publicKey: CryptoKey; privateKey: CryptoKey },
  ): Promise<RatchetSession> {
    const dhPub = await exportPublicKeyRaw(localDhKeyPair.publicKey);
    const dhPriv = await exportPrivateKeyPkcs8(localDhKeyPair.privateKey);

    return {
      rootKey: sharedSecret,
      chainKeySend: new Uint8Array(32),
      chainKeyRecv: new Uint8Array(32),
      dhSendPub: dhPub,
      dhSendPriv: dhPriv,
      dhRecvPub: new Uint8Array(32),
      nSend: 0,
      nRecv: 0,
      pn: 0,
      skippedMessageKeys: new Map(),
    };
  }

  /**
   * Encrypt a message, advancing the sending chain.
   */
  static async encrypt(
    session: RatchetSession,
    plaintext: Uint8Array,
  ): Promise<EncryptResult> {
    const [chainKey, messageKey] = await chainKdf(session.chainKeySend);

    const ciphertext = await aesGcmEncrypt(messageKey, plaintext);

    const header: RatchetHeader = {
      type: 'double_ratchet',
      dh: session.dhSendPub,
      n: session.nSend,
      pn: session.pn,
    };

    const nextSession: RatchetSession = {
      ...session,
      chainKeySend: chainKey,
      nSend: session.nSend + 1,
      skippedMessageKeys: new Map(session.skippedMessageKeys),
    };

    return { ciphertext, header, nextSession };
  }

  /**
   * Decrypt a message, advancing the receiving chain.
   */
  static async decrypt(
    session: RatchetSession,
    ciphertext: Uint8Array,
    header: RatchetHeader,
  ): Promise<DecryptResult> {
    // Check skipped message keys first
    const skippedKey = `${base64urlEncode(header.dh)}:${header.n}`;
    if (session.skippedMessageKeys.has(skippedKey)) {
      const messageKey = session.skippedMessageKeys.get(skippedKey)!;
      const nextSkipped = new Map(session.skippedMessageKeys);
      nextSkipped.delete(skippedKey);
      try {
        const plaintext = await aesGcmDecrypt(messageKey, ciphertext);
        return {
          plaintext,
          nextSession: { ...session, skippedMessageKeys: nextSkipped },
        };
      } catch (err) {
        throw SdkError.decryption('Failed to decrypt with skipped message key', err);
      }
    }

    let currentSession = { ...session, skippedMessageKeys: new Map(session.skippedMessageKeys) };

    // DH ratchet step if new sender DH key
    const dhHeaderEnc = base64urlEncode(header.dh);
    const dhRecvEnc = base64urlEncode(currentSession.dhRecvPub);

    if (dhHeaderEnc !== dhRecvEnc) {
      // Skip ahead in current receiving chain (pn messages)
      currentSession = await skipMessageKeys(currentSession, header.pn);
      // Do DH ratchet step
      currentSession = await dhRatchetStep(currentSession, header.dh);
    }

    // Skip ahead to message n
    currentSession = await skipMessageKeys(currentSession, header.n);

    // Derive message key
    const [nextChainKey, messageKey] = await chainKdf(currentSession.chainKeyRecv);
    currentSession = {
      ...currentSession,
      chainKeyRecv: nextChainKey,
      nRecv: currentSession.nRecv + 1,
    };

    try {
      const plaintext = await aesGcmDecrypt(messageKey, ciphertext);
      return { plaintext, nextSession: currentSession };
    } catch (err) {
      throw SdkError.decryption('AES-GCM decryption failed', err);
    }
  }

  /**
   * Serialize RatchetSession to storable RatchetState.
   */
  static serialize(session: RatchetSession, conversationId: string): RatchetState {
    const skippedMessageKeys: Record<string, string> = {};
    for (const [k, v] of session.skippedMessageKeys) {
      skippedMessageKeys[k] = base64urlEncode(v);
    }
    return {
      conversationId,
      rootKey: base64urlEncode(session.rootKey),
      chainKeySend: base64urlEncode(session.chainKeySend),
      chainKeyRecv: base64urlEncode(session.chainKeyRecv),
      dhSendPub: base64urlEncode(session.dhSendPub),
      dhSendPriv: base64urlEncode(session.dhSendPriv),
      dhRecvPub: base64urlEncode(session.dhRecvPub),
      nSend: session.nSend,
      nRecv: session.nRecv,
      pn: session.pn,
      skippedMessageKeys,
    };
  }

  /**
   * Deserialize a stored RatchetState to a RatchetSession.
   */
  static deserialize(state: RatchetState): RatchetSession {
    const skippedMessageKeys = new Map<string, Uint8Array>();
    for (const [k, v] of Object.entries(state.skippedMessageKeys)) {
      skippedMessageKeys.set(k, base64urlDecode(v));
    }
    return {
      rootKey: base64urlDecode(state.rootKey),
      chainKeySend: base64urlDecode(state.chainKeySend),
      chainKeyRecv: base64urlDecode(state.chainKeyRecv),
      dhSendPub: base64urlDecode(state.dhSendPub),
      dhSendPriv: base64urlDecode(state.dhSendPriv),
      dhRecvPub: base64urlDecode(state.dhRecvPub),
      nSend: state.nSend,
      nRecv: state.nRecv,
      pn: state.pn,
      skippedMessageKeys,
    };
  }
}

// --- Helpers ---

async function ratchetKdf(
  rootKey: Uint8Array,
  dhOutput: Uint8Array,
): Promise<[Uint8Array, Uint8Array]> {
  const output = await hkdf(dhOutput, HKDF_INFO.RATCHET_ROOT, 64, rootKey);
  return [output.slice(0, 32), output.slice(32, 64)];
}

async function chainKdf(chainKey: Uint8Array): Promise<[Uint8Array, Uint8Array]> {
  const [nextChain, msgKey] = await Promise.all([
    hkdf(chainKey, HKDF_INFO.RATCHET_CHAIN, 32),
    hkdf(chainKey, HKDF_INFO.MESSAGE_KEY, 32),
  ]);
  return [nextChain, msgKey];
}

async function skipMessageKeys(
  session: RatchetSession,
  until: number,
): Promise<RatchetSession> {
  if (session.nRecv >= until) return session;

  const totalSkip = until - session.nRecv;
  if (session.skippedMessageKeys.size + totalSkip > MAX_SKIP) {
    throw new SdkError(SdkErrorCode.DECRYPTION_ERROR, 'Too many skipped messages');
  }

  const skipped = new Map(session.skippedMessageKeys);
  let chainKey = session.chainKeyRecv;
  const dhKey = base64urlEncode(session.dhRecvPub);
  let n = session.nRecv;

  while (n < until) {
    const [nextChain, messageKey] = await chainKdf(chainKey);
    skipped.set(`${dhKey}:${n}`, messageKey);
    chainKey = nextChain;
    n++;
  }

  return { ...session, chainKeyRecv: chainKey, nRecv: n, skippedMessageKeys: skipped };
}

async function dhRatchetStep(
  session: RatchetSession,
  newDhRecvPub: Uint8Array,
): Promise<RatchetSession> {
  // Generate new DH key pair
  const newDhKp = await generateX25519KeyPair();
  const newDhPub = await exportPublicKeyRaw(newDhKp.publicKey);
  const newDhPriv = await exportPrivateKeyPkcs8(newDhKp.privateKey);

  const recipientPub = await importX25519PublicKey(newDhRecvPub);

  // Receiving chain
  const dhRecv = await x25519Dh(await importX25519PrivateKey(session.dhSendPriv), recipientPub);
  const [rootKey1, chainKeyRecv] = await ratchetKdf(session.rootKey, dhRecv);

  // Sending chain
  const dhSend = await x25519Dh(newDhKp.privateKey, recipientPub);
  const [rootKey2, chainKeySend] = await ratchetKdf(rootKey1, dhSend);

  return {
    ...session,
    rootKey: rootKey2,
    chainKeySend,
    chainKeyRecv,
    dhSendPub: newDhPub,
    dhSendPriv: newDhPriv,
    dhRecvPub: newDhRecvPub,
    nSend: 0,
    nRecv: 0,
    pn: session.nSend,
  };
}

File: flutter_sdk/lib/src/crypto/ratchet_engine.dart
import 'dart:typed_data';
import 'package:cryptography/cryptography.dart';
import 'crypto_types.dart';
import '../session/session_data.dart';
import '../errors/sdk_error.dart';

/// Maximum number of skipped message keys stored per session.
const _maxSkippedKeys = 2000;

/// Maximum number of keys to skip in a single ratchet step.
const _maxSkip = 1000;

/// Implements the Double Ratchet Algorithm for 1:1 end-to-end encryption.
///
/// Provides forward secrecy and break-in recovery. Each message is encrypted
/// with a unique key derived from the ratchet chain.
class RatchetEngine {
  static final _x25519 = X25519();
  static final _aesGcm = AesGcm.with256bits();

  static const _rootKdfInfo = 'rust_e2e_root_ratchet';
  static const _chainKdfInfo = 'rust_e2e_chain_key';
  static const _messageKdfInfo = 'rust_e2e_message_key';

  /// Initialize the Double Ratchet session from the sender side.
  ///
  /// [sharedSecret] is the X3DH output.
  /// [recipientDhPublicKey] is the recipient's signed pre-key public bytes.
  static Future<RatchetState> initSender(
    Uint8List sharedSecret,
    Uint8List recipientDhPublicKey,
    String conversationId,
  ) async {
    // Generate the initial DH ratchet key pair for sending
    final dhPair = await _generateDhPair();

    // Derive root and chain keys
    final dhOut = await _dh(dhPair.privateKey, recipientDhPublicKey);
    final (rootKey, chainKeySend) =
        await _kdfRootKey(sharedSecret, dhOut);

    return RatchetState(
      conversationId: conversationId,
      rootKey: rootKey,
      chainKeySend: chainKeySend,
      chainKeyRecv: Uint8List(32), // empty until first message received
      dhSendPub: dhPair.publicKey,
      dhSendPriv: dhPair.privateKey,
      dhRecvPub: recipientDhPublicKey,
      nSend: 0,
      nRecv: 0,
      pn: 0,
      skippedMessageKeys: {},
    );
  }

  /// Initialize the Double Ratchet session from the receiver side.
  ///
  /// [sharedSecret] is the X3DH output.
  /// [localDhPair] is the receiver's signed pre-key pair.
  static RatchetState initReceiver(
    Uint8List sharedSecret,
    Curve25519KeyPair localDhPair,
    String conversationId,
  ) {
    return RatchetState(
      conversationId: conversationId,
      rootKey: sharedSecret,
      chainKeySend: Uint8List(32),
      chainKeyRecv: Uint8List(32),
      dhSendPub: localDhPair.publicKey,
      dhSendPriv: localDhPair.privateKey,
      dhRecvPub: Uint8List(32), // populated on first received message
      nSend: 0,
      nRecv: 0,
      pn: 0,
      skippedMessageKeys: {},
    );
  }

  /// Encrypts [plaintext] using the current ratchet [state].
  ///
  /// Returns the updated [RatchetState] and the [RatchetCiphertext].
  static Future<(RatchetState, RatchetCiphertext)> encrypt(
    RatchetState state,
    Uint8List plaintext, {
    bool isInitialMessage = false,
    X3dhInitHeader? x3dhHeader,
  }) async {
    final (newChainKey, messageKey) = await _kdfChainKey(state.chainKeySend);
    final ciphertext = await _encryptMessage(messageKey, plaintext);

    final header = RatchetMessageHeader(
      dhPublicKey: state.dhSendPub,
      messageNumber: state.nSend,
      previousChainLength: state.pn,
      isInitialMessage: isInitialMessage,
      x3dhHeader: x3dhHeader,
    );

    final newState = state.copyWith(
      chainKeySend: newChainKey,
      nSend: state.nSend + 1,
    );

    return (newState, RatchetCiphertext(ciphertext: ciphertext, header: header));
  }

  /// Decrypts [ciphertext] using the current ratchet [state].
  ///
  /// Returns the updated [RatchetState] and the decrypted plaintext bytes.
  /// Throws [DecryptionError] if decryption fails.
  static Future<(RatchetState, Uint8List)> decrypt(
    RatchetState state,
    RatchetCiphertext ciphertext,
  ) async {
    final header = ciphertext.header;

    // Check if this is a skipped message key
    final skipKey = _skippedKey(state, header);
    if (skipKey != null) {
      final plaintext = await _decryptMessage(skipKey, ciphertext.ciphertext)
          .then((p) => p)
          .catchError((_) => throw DecryptionError(
              conversationId: state.conversationId, seq: header.messageNumber));
      final newSkipped = Map<int, Uint8List>.from(state.skippedMessageKeys)
        ..remove(header.messageNumber);
      return (state.copyWith(skippedMessageKeys: newSkipped), plaintext);
    }

    RatchetState currentState = state;

    // If this message is from a new DH ratchet step
    final isNewDhKey = !_bytesEqual(header.dhPublicKey, state.dhRecvPub);

    if (isNewDhKey) {
      // Skip messages from the old chain
      currentState = await _skipMessageKeys(currentState, header.previousChainLength);
      // Perform DH ratchet step
      currentState = await _dhRatchetStep(currentState, header.dhPublicKey);
    }

    // Skip messages in the current receiving chain
    currentState =
        await _skipMessageKeys(currentState, header.messageNumber);

    // Derive message key
    final (newChainKeyRecv, messageKey) =
        await _kdfChainKey(currentState.chainKeyRecv);
    currentState = currentState.copyWith(
      chainKeyRecv: newChainKeyRecv,
      nRecv: currentState.nRecv + 1,
    );

    try {
      final plaintext = await _decryptMessage(messageKey, ciphertext.ciphertext);
      return (currentState, plaintext);
    } catch (_) {
      throw DecryptionError(
          conversationId: state.conversationId, seq: header.messageNumber);
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  static Uint8List? _skippedKey(RatchetState state, RatchetMessageHeader header) {
    if (_bytesEqual(header.dhPublicKey, state.dhRecvPub)) {
      return state.skippedMessageKeys[header.messageNumber];
    }
    return null;
  }

  static Future<RatchetState> _skipMessageKeys(
      RatchetState state, int until) async {
    if (state.nRecv > until) return state;
    if (until - state.nRecv > _maxSkip) {
      throw const StorageError(reason: 'Too many skipped messages');
    }

    var currentState = state;
    final skipped = Map<int, Uint8List>.from(state.skippedMessageKeys);

    while (currentState.nRecv < until) {
      final (newChain, msgKey) = await _kdfChainKey(currentState.chainKeyRecv);
      skipped[currentState.nRecv] = msgKey;
      currentState = currentState.copyWith(
        chainKeyRecv: newChain,
        nRecv: currentState.nRecv + 1,
      );

      // Cap skipped message keys
      if (skipped.length > _maxSkippedKeys) {
        final oldest = skipped.keys.reduce((a, b) => a < b ? a : b);
        skipped.remove(oldest);
      }
    }

    return currentState.copyWith(skippedMessageKeys: skipped);
  }

  static Future<RatchetState> _dhRatchetStep(
      RatchetState state, Uint8List remotePublicKey) async {
    // Derive new recv chain key
    final dhRecv = await _dh(state.dhSendPriv, remotePublicKey);
    final (rootKey1, chainKeyRecv) = await _kdfRootKey(state.rootKey, dhRecv);

    // Generate new DH key pair for sending
    final newDhPair = await _generateDhPair();

    // Derive new send chain key
    final dhSend = await _dh(newDhPair.privateKey, remotePublicKey);
    final (rootKey2, chainKeySend) = await _kdfRootKey(rootKey1, dhSend);

    return state.copyWith(
      rootKey: rootKey2,
      chainKeySend: chainKeySend,
      chainKeyRecv: chainKeyRecv,
      dhSendPub: newDhPair.publicKey,
      dhSendPriv: newDhPair.privateKey,
      dhRecvPub: remotePublicKey,
      nSend: 0,
      nRecv: 0,
      pn: state.nSend,
    );
  }

  static Future<(Uint8List rootKey, Uint8List chainKey)> _kdfRootKey(
      Uint8List rootKey, Uint8List dhOutput) async {
    final hkdf = Hkdf(hmac: Hmac.sha256(), outputLength: 64);
    final output = await hkdf.deriveKey(
      secretKey: SecretKey(dhOutput),
      nonce: rootKey,
      info: _rootKdfInfo.codeUnits,
    );
    final bytes = await output.extractBytes();
    return (
      Uint8List.fromList(bytes.sublist(0, 32)),
      Uint8List.fromList(bytes.sublist(32, 64)),
    );
  }

  static Future<(Uint8List chainKey, Uint8List messageKey)> _kdfChainKey(
      Uint8List chainKey) async {
    final hkdf32 = Hkdf(hmac: Hmac.sha256(), outputLength: 32);
    final newChain = await hkdf32.deriveKey(
      secretKey: SecretKey(chainKey),
      nonce: Uint8List.fromList([0x01]),
      info: _chainKdfInfo.codeUnits,
    );
    final msgKey = await hkdf32.deriveKey(
      secretKey: SecretKey(chainKey),
      nonce: Uint8List.fromList([0x02]),
      info: _messageKdfInfo.codeUnits,
    );
    return (
      Uint8List.fromList(await newChain.extractBytes()),
      Uint8List.fromList(await msgKey.extractBytes()),
    );
  }

  static Future<Uint8List> _encryptMessage(
      Uint8List key, Uint8List plaintext) async {
    final secretKey = SecretKey(key);
    final nonce = _aesGcm.newNonce();
    final box = await _aesGcm.encrypt(plaintext,
        secretKey: secretKey, nonce: nonce);
    // Prepend nonce (12 bytes) to ciphertext
    final result = Uint8List(12 + box.cipherText.length + box.mac.bytes.length);
    result.setRange(0, 12, nonce);
    result.setRange(12, 12 + box.cipherText.length, box.cipherText);
    result.setRange(12 + box.cipherText.length, result.length, box.mac.bytes);
    return result;
  }

  static Future<Uint8List> _decryptMessage(
      Uint8List key, Uint8List ciphertextWithNonce) async {
    if (ciphertextWithNonce.length < 28) {
      throw const StorageError(reason: 'Ciphertext too short');
    }
    final nonce = ciphertextWithNonce.sublist(0, 12);
    final mac = ciphertextWithNonce.sublist(ciphertextWithNonce.length - 16);
    final ciphertext =
        ciphertextWithNonce.sublist(12, ciphertextWithNonce.length - 16);

    final secretKey = SecretKey(key);
    final box = SecretBox(ciphertext, nonce: nonce, mac: Mac(mac));
    final plaintext = await _aesGcm.decrypt(box, secretKey: secretKey);
    return Uint8List.fromList(plaintext);
  }

  static Future<Uint8List> _dh(
      Uint8List privateKeyBytes, Uint8List remotePublicKeyBytes) async {
    final privateKey = await _x25519.newKeyPairFromSeed(privateKeyBytes);
    final remotePublicKey =
        SimplePublicKey(remotePublicKeyBytes, type: KeyPairType.x25519);
    final sharedSecret = await _x25519.sharedSecretKey(
        keyPair: privateKey, remotePublicKey: remotePublicKey);
    return Uint8List.fromList(await sharedSecret.extractBytes());
  }

  static Future<Curve25519KeyPair> _generateDhPair() async {
    final keyPair = await _x25519.newKeyPair();
    final privateBytes = await keyPair.extractPrivateKeyBytes();
    final publicKey = await keyPair.extractPublicKey();
    return Curve25519KeyPair(
      publicKey: Uint8List.fromList(publicKey.bytes),
      privateKey: Uint8List.fromList(privateBytes),
    );
  }

  static bool _bytesEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}
